htslib/sam.h

```diff
@@ -1090,6 +1090,73 @@ hts_pos_t bam_cigar2rlen(int n_cigar, const uint32_t *cigar);
 HTSLIB_EXPORT
 hts_pos_t bam_endpos(const bam1_t *b);
 
+/*! @typedef
+ *  @abstract Flag filtering settings structure
+ */
+typedef struct sam_flag_filter_s {
+    uint32_t has_all:1, has_none:1, has_any:1, has_omit_any:1, has_exact:1, :27;
+    // To pass this filter, a read must have...
+    unsigned int all;       // all of these flags
+    unsigned int none;      // none of these
+    unsigned int any;       // at least one of these
+    unsigned int omit_any;  // OMIT at least one of these
+    unsigned int exact;     // exactly these flags (and no others)
+} sam_flag_filter_t;
+
+/// Check whether the flags pass the collected flag filters
+/** @param flag  Bitwise SAM flags
+    @param filt  Pointer to the flag filtering settings
+    @return      1 if @a flag passes all the active filters, 0 if not
+*/
+static inline
+int sam_flag_passes_flag_filter(unsigned int flag, const sam_flag_filter_t *filt) {
+    if (filt->has_all && (flag & filt->all) != filt->all) return 0;
+    if (filt->has_none && (flag & filt->none) != 0) return 0;
+    if (filt->has_any && (flag & filt->any) == 0) return 0;
+    if (filt->has_omit_any && (flag & filt->omit_any) == filt->omit_any) return 0;
+    if (filt->has_exact && flag != filt->exact) return 0;
```

Comment:
Yes, I've been thinking about reworking it so `exact:` is implemented in terms of those two as well. The downside is if someone specifies all three it will discard some of their settings, but no-one should be using `exact:` in conjunction with the others anyway…